File: Windows/Shiba/Shiba.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Shiba.CommonProperty;
using Shiba.Controls;
using Shiba.ExtensionExecutors;
using Shiba.Scripting;
using Shiba.Scripting.Runtime;

namespace Shiba
{
    public class ShibaApp
    {
        private ShibaApp(Action<ShibaConfiguration> action)
        {
            action?.Invoke(Configuration);
            ViewMapping.Init();
        }

        public ViewMapping ViewMapping { get; } = new ViewMapping();

        public ShibaConfiguration Configuration { get; } = new ShibaConfiguration();

        public static ShibaApp Instance { get; protected set; }

        public void AddConverter(string converter)
        {
            Configuration.ScriptRuntime.Execute(converter);
        }

        public static void Init(Action<ShibaConfiguration> action = null)
        {
            Instance = new ShibaApp(c =>
            {
                c.PlatformType = "UWP";
                c.ScriptRuntime.AddObject("storage", new Storage());
                action?.Invoke(c);
            });
        }
    }

    public class ShibaConfiguration
    {
        public IScriptRuntime ScriptRuntime { get; set; } = new DefaultScriptRuntime();
        public string PlatformType { get; set; } = "Windows";

        public List<IExtensionExecutor> ExtensionExecutors { get; } =
            AppDomain.CurrentDomain.GetAssemblies()
                .Where(it => !it.IsDynamic)
                .SelectMany(it => it.ExportedTypes)
                .Where(it => it.IsClass && !it.IsAbstract && typeof(IExtensionExecutor).IsAssignableFrom(it))
                .Select(it => Activator.CreateInstance(it) as IExtensionExecutor).ToList();

        public List<ICommonProperty> CommonProperties { get; } =
            AppDomain.CurrentDomain.GetAssemblies()
                .Where(it => !it.IsDynamic)
                .SelectMany(it => it.ExportedTypes)
                .Where(it => it.IsClass && !it.IsAbstract && typeof(ICommonProperty).IsAssignableFrom(it))
                .Select(it => Activator.CreateInstance(it) as ICommonProperty).ToList();
    }
}

File: Android/shiba/src/main/java/moe/tlaster/shiba/Shiba.kt
package moe.tlaster.shiba

import android.content.Context
import android.util.ArrayMap
import android.util.Log
import moe.tlaster.shiba.commonProperty.GridProperty
import moe.tlaster.shiba.commonProperty.ICommonProperty
import moe.tlaster.shiba.extensionExecutor.BindingExecutor
import moe.tlaster.shiba.extensionExecutor.IExtensionExecutor
import moe.tlaster.shiba.extensionExecutor.JsonExecutor
import moe.tlaster.shiba.mapper.*
import moe.tlaster.shiba.scripting.DefaultScriptRuntime
import moe.tlaster.shiba.scripting.IScriptRuntime
import moe.tlaster.shiba.scripting.runtime.Http
import moe.tlaster.shiba.scripting.runtime.Storage
import moe.tlaster.shiba.type.View
import org.liquidplayer.javascript.JSFunction
import org.liquidplayer.javascript.JSObject

internal typealias NativeView = android.view.View
internal typealias ShibaView = View

class ShibaConfiguration {
    var scriptRuntime: IScriptRuntime = DefaultScriptRuntime()
    var platformType = "Android"
    val commonProperties = ArrayList<ICommonProperty>()
    val extensionExecutors = ArrayList<IExtensionExecutor>()
    val nativeConverter = androidx.collection.ArrayMap<String, ((List<Any?>) -> Any?)>()
}

object Shiba {
    internal var appComponent: ShibaView? = null
    val viewMapping = ArrayMap<String, IViewMapper<*>>()
    val configuration = ShibaConfiguration()
    internal val components = ArrayMap<String, ShibaView>()

    fun init(application: Context) {
        addRenderer("stack", StackMapper())
        addRenderer("text", TextMapper())
        addRenderer("input", InputMapper())
        addRenderer("list", ListMapper())
        addRenderer("grid", GridMapper())
        addRenderer("button", ButtonMapper())
        addExtensionExecutor(BindingExecutor())
        addExtensionExecutor(JsonExecutor())
        configuration.commonProperties.add(GridProperty())
        if (configuration.scriptRuntime is DefaultScriptRuntime) {
            (configuration.scriptRuntime as DefaultScriptRuntime).addObject("shibaStorage") {
                Storage(application, it)
            }
            (configuration.scriptRuntime as DefaultScriptRuntime).addObject("http") {
                Http(it)
            }
            (configuration.scriptRuntime as DefaultScriptRuntime).addObject("console") {
                object : JSObject(it) {
                    init {
                        property("log", object : JSFunction(it, "log") {
                            fun log(msg: String) {
                                Log.i("Javascript", msg)
                            }
                        })
                    }
                }
            }
        }
    }

    fun addConverter(name: String, converter: ((List<Any?>) -> Any?)) {
        configuration.nativeConverter[name] = converter
    }

    fun addRenderer(name: String, mapper: IViewMapper<*>) {
        viewMapping[name] = mapper
    }

    fun addExtensionExecutor(executor: IExtensionExecutor) {
        configuration.extensionExecutors.add(executor)
    }
}